### PR TITLE
Extract detecting CC/CXX into helper script so we can use it from mono.proj

### PR DESCRIPTION
**Pull request by akoeplinger**
Also fixed a small typo in the detection script, see first commit.

Closes https://github.com/dotnet/runtime/pull/31771

### CODE REVIEW

**Comment by am11** [comment]
Sorry about that typo, I happened to stumble upon that condition today on different platform and fixed in https://github.com/dotnet/runtime/pull/31865 😄 (will revert from there).
+1 for reusing the code.

File: eng/native/detect-compiler.sh [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env bash
```
Comment (am11) [inline eng/native/detect-compiler.sh]: nit: we can call it `init-compiler.sh`, to rhyme with the existing `init-distro-rid.sh` and upcoming `init-os-and-arch.sh` (from https://github.com/dotnet/runtime/pull/31701)?

Comment (akoeplinger) [reply eng/native/detect-compiler.sh]: yup sounds good

**Comment by davidwrighton** [comment]
@akoeplinger This works. It fixes the issue I hit nicely. I'm going to close my PR as this supersedes it.

**Review by EgorBo — APPROVED** [review]

File: src/mono/mono.proj [hunk 2]
```diff
@@ -31,13 +31,13 @@
...
         <_MonoConfigureParams Condition="$(MonoEnableLLVM) == true">$(_MonoConfigureParams) --enable--llvm </_MonoConfigureParams>
-        <_MonoConfigureParams>$(_MonoConfigureParams) --with-core=only CC=clang CXX=clang++ CFLAGS="$(_MonoExtraCFLAGS)" CXXFLAGS="$(_MonoExtraCXXFLAGS)"</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) --enable-maintainer-mode --enable-compile-warnings --with-core=only CFLAGS="$(_MonoExtraCFLAGS)" CXXFLAGS="$(_MonoExtraCXXFLAGS)"</_MonoConfigureParams>
     </PropertyGroup>
 
     <Message Text="Configuring Mono with '$(_MonoConfigureParams)'" Importance="High" />
     <MakeDir Directories="$(MonoObjDir)" />
     <Exec Command="NOCONFIGURE=1 $(MonoProjectRoot)autogen.sh" />
-    <Exec Command="$(MonoProjectRoot)configure $(_MonoConfigureParams)" WorkingDirectory="$(MonoObjDir)" />
+    <Exec Command="bash -c 'source $(RepositoryEngineeringDir)native/init-compiler.sh $(Platform) clang &amp;&amp; $(MonoProjectRoot)configure $(_MonoConfigureParams)'" WorkingDirectory="$(MonoObjDir)" />
```
Comment (am11) [inline src/mono/mono.proj:40]: @akoeplinger, would it be possible to let `eng/build.sh` determine which compiler to use? i.e. the hardcoded `clang` to be replaced by `$(Compiler)` defined here: https://github.com/dotnet/runtime/blob/2ef008e29b7f52d6dee8be3ea0bb675651cccb8f/eng/build.sh#L275-L292
Doing that will help systems which require gcc build (`runtime/build.sh -gcc ...`).

Comment (akoeplinger) [reply src/mono/mono.proj:40]: @am11 sure, that would be great!

**Pull request merged**